webpack/scenes/ContentViews/Details/Promote/ContentViewVersionPromote.js

```diff
@@ -0,0 +1,165 @@
+import Loading from '../../../../components/Loading';
+
+const ContentViewVersionPromote = ({
+  cvId, versionIdToPromote, versionNameToPromote, versionEnvironments, show, setIsOpen,
+}) => {
+  const [description, setDescription] = useState('');
+  const [userCheckedItems, setUserCheckedItems] = useState([]);
+  const [saving, setSaving] = useState(false);
+  const [forcePromote, setForcePromote] = useState([]);
+  const environmentPathResponse = useSelector(selectEnvironmentPaths);
+  const environmentPathStatus = useSelector(selectEnvironmentPathsStatus);
+  const environmentPathLoading = environmentPathStatus === STATUS.PENDING;
+  const promoteResponse = useSelector(state =>
+    selectPromoteCVVersionResponse(state, versionIdToPromote, versionEnvironments));
+  const promoteStatus = useSelector(state =>
+    selectPromoteCVVersionStatus(state, versionIdToPromote, versionEnvironments));
+  const promoteError = useSelector(state =>
+    selectPromoteCVVersionError(state, versionIdToPromote, versionEnvironments));
+  const promoteResolved = promoteStatus === STATUS.RESOLVED;
+  const dispatch = useDispatch();
+
+  const onPromote = () => {
+    setSaving(true);
+    dispatch(promoteContentViewVersion({
+      id: versionIdToPromote,
+      description,
+      versionEnvironments,
+      environment_ids: userCheckedItems.map(item => item.id),
+      force: true,
+    }));
+  };
+
+  useEffect(() => {
+    if (promoteResolved && promoteResponse) {
+      setIsOpen(false);
+      dispatch(getContentViewVersions(cvId));
+    }
+    if (promoteError) {
+      setSaving(false);
+    }
+  }, [promoteResponse, promoteResolved, promoteError, setSaving, setIsOpen, dispatch, cvId]);
+
+  const envPathFlat = useMemo(() => {
+    if (!environmentPathLoading) {
+      const { results } = environmentPathResponse || {};
+      return results.map(result => result.environments).flatten();
+    }
+    return [];
+  }, [environmentPathResponse, environmentPathLoading]);
+
+  const prior = useCallback(
+    env => envPathFlat.find(item => item.id === env.prior.id),
+    [envPathFlat],
+  );
+
+  const isChecked = useCallback(
+    env => (userCheckedItems.includes(env) ||
+      versionEnvironments.filter(item => item.id === env.id).length),
+    [userCheckedItems, versionEnvironments],
+  );
+
+  const isValid = useCallback((env) => {
+    if (!env.prior) return true;
+    if (!isChecked(prior(env))) return false;
+    return isValid(prior(env));
+  }, [prior, isChecked]);
+
+  useEffect(() => {
+    setForcePromote(userCheckedItems.filter(item => !isValid(item)));
+  }, [userCheckedItems, setForcePromote, isValid]);
+
+  return (
+    <Modal
+      title={__(`Promote version ${versionNameToPromote}`)}
+      isOpen={show}
+      variant={ModalVariant.large}
+      onClose={() => {
+        setIsOpen(false);
+      }}
+      appendTo={document.body}
+    >
+      {saving && <Loading loadingText={__('Please wait while task starts..')} />}
```

Comment:
Would naming this variable 'promoting' instead of 'saving' be more clear or accurate?